Site: Document animations. #2532 (#2580)
sveltejs/svelte: site/content/docs/02-template-syntax.md

### Animations

```sv
animate:name
```

```sv
animate:name={params}
```

```js
animation = (node: HTMLElement, { from: DOMRect, to: DOMRect } , params: any) => {
	delay?: number,
	duration?: number,
	easing?: (t: number) => number,
	css?: (t: number, u: number) => string,
	tick?: (t: number, u: number) => void
}
```

```js
DOMRect {
	bottom: number,
	height: number,
	​​left: number,
	right: number,
	​top: number,
	width: number,
	x: number,
	y:number
}
```

---

> An animation is triggered when the contents of a [keyed each block](docs#Each_blocks) are re-ordered. Animations do not run when an element is removed only when the each block's data is reordered. Animate directives must be on an element that is an *immediate* child of a keyed each block.

Reviewer: add a comma after "removed"

> An animation is triggered when the contents of a [keyed each block](docs#Each_blocks) are re-ordered. Animations do not run when an element is removed, only when the each block's data is reordered. Animate directives must be on an element that is an *immediate* child of a keyed each block.

Reviewer 2: `An animation is triggered when the elements of a [keyed each block](docs#Each_blocks) are re-ordered. Animations do not run for elements that are added or removed (like transitions are), only for those that are reordered. Animate directives must be on an element that is an *immediate* child of a keyed each block.`

Maybe something like this? I'm not thrilled about the parenthetical (like transitions are) shoehorned in there. I'm on the lookout for a better way to work that into the paragraph.

Author: Yeah, this works better. I think I made the assumption when writing this that people reading about `animate` would know about `transition` but each entry should stand alone.

> Animations will normally be used with the [`flip`](docs#flip) function which will swap the positions of elements when their order changes.

Reviewer: missing link

Author: It is currently missing but I felt it would be better to add it now because I'm writing the docs for `flip` as we speak. Otherwise we'd have to PR again just to add the link.

> Animations will normally be used with the [`flip`](docs#flip) function which will swap the positions of elements when their order changes.

Reviewer 2: I don't feel like saying they're normally used with `flip` is very helpful. Or at least I think I'd find it a little misleading. I'd probably prefer this whole thing if it made it clearer that `flip` is just a built-in animation that ships with Svelte.

Seeing as this is reference documentation, there may also be an argument for just introducing the general concept of a custom animation function here, and leaving talking about `flip` for its own section below. This isn't a tutorial, we aren't necessarily trying to ease (editor's note: ba-dum-psh) someone into the concept.

Author: This is a good point. My original reasoning for putting it in is that this is the only documentation we have aside from the interactive tutorial. 

Perhaps a mention that the directive can be used with custom animations or Svelte's built-in animations and relevant links rather than any guidance?

```html
<!-- When `list` is reordered the animation will run-->
{#each list as item, index (item)}
	<li animate:flip>{item}</li>
{/each}
```

#### Animation Parameters

---

As with actions and transitions, animations can have parameters.

(The double `{{curlies}}` aren't a special syntax; this is an object literal inside an expression tag.)

```html
{#each list as item, index (item)}
	<li animate:flip="{{ delay: 500 }}">{item}</li>
{/each}
```

#### Custom animation functions

---

> Animations can use custom functions, these are very similar to custom transition functions. Custom animations also give you access to an `animation` parameter which is an object containing `from` and `to` properties containing the [DOMRect](https://developer.mozilla.org/en-US/docs/Web/API/DOMRect#Properties) (geometry) of the element. The `from` property is the geometry of the element in its starting position, the `to` property is the geometry of the element after the list has been re-ordered and the DOM updated.

Reviewer 2: Maybe make it clearer here somehow that `from` and `to` are each separate instances of `DOMRect`?

> If the returned object has a css function, Svelte will create a CSS animation that plays on the element.

Reviewer 2: ```suggestion
If the returned object has a `css` method, Svelte will create a CSS animation that plays on the element.
```

> The `t` argument passed to css is a value between 0 and 1 after the easing function has been applied. `t` always starts with a value of 0 and ends with a value of 1. `u` always starts with a value of 1 and ends with a value of 0.

Reviewer 2: If this was copy-pasted from the transitions section, it lost some formatting. I'm not sure it's necessary to specifically mention that ranges of both t and u.

Maybe:

```suggestion
The `t` argument passed to `css` is a value that goes from `0` and `1` after the `easing` function has been applied. The `u` argument is equal to `1 - t`.
```

> The function is called repeatedly before the transition begins, with different `t` and `u` arguments.

Reviewer 2: ```suggestion
 The function is called repeatedly *before* the transition begins, with different `t` and `u` arguments.
```

To make this match the corresponding sentence in the transitions section.

```html
<script>
	import { flip } from 'svelte/animate';
	import { cubicOut } from 'svelte/easing';

	function whizz(node, { from, to }, params) {

		const dx = from.left - to.left;
		const dy = from.top - to.top;

		const d = Math.sqrt(dx * dx + dy * dy);

		return {
			delay: 0,
			duration: Math.sqrt(d) * 120,
			easing: cubicOut,
			css: (t, u) =>
				`transform: translate(${u * dx}px, ${u * dy}px) rotate(${t*360}deg);`
		};
	}
</script>

{#each list as item, index (item)}
	<div animate:whizz>{item}</div>
{/each}
```

---

> A custom transition function can also return a tick function, which is called during the transition with the same `t` and `u` arguments.

Reviewer 2: ```suggestion
A custom animation function can also return a `tick` function, which is called *during* the animation with the same `t` and `u` arguments.
```

> If it's possible to use `css` instead of `tick`, do so — CSS animations can run off the main thread, preventing jank on slower devices.

```html
<script>

> 	import { flip } from 'svelte/animate';

Reviewer 2: This is a custom animation, it doesn't use `flip`.
